Add technical docs on diagnosing instability etc (#85074)
elastic/elasticsearch: docs/reference/modules/network/threading.asciidoc

[[modules-network-threading-model]]
==== Networking threading model

> This section describes some technical details about the threading model used by
> the networking subsystem in {es}, which may be important for advanced users to
> diagnose certain problems in a cluster.

Reviewer: ```suggestion
This section describes the threading model used by the networking subsystem in
{es}. While this information isn't required to use {es}, it may be useful to advanced
users who are diagnosing network problems in a cluster.
```

{es} nodes communicate over a collection of TCP channels that together form a
transport connection. {es} clients communicate with the cluster over HTTP,
which also uses one or more TCP channels. Each of these TCP channels is owned

> by exactly one of the `transport_worker` threads in the node. The owning thread
> of a channel is chosen when the channel is opened, and cannot change.

Reviewer: ```suggestion
by exactly one of the `transport_worker` threads in the node. This owning thread
is chosen when the channel is opened and cannot change.
```

Each `transport_worker` thread has sole responsibility for sending and
receiving data over the channels it owns. One of the `transport_worker` threads

> is also responsible for accepting new incoming transport connections, and one
> is responsible for accepting new HTTP connections.

Reviewer: ```suggestion
is also responsible for accepting new incoming transport connections. Another
`transport_worker` thread is responsible for accepting new HTTP connections.
```

Author: I know what you mean but it might be the same thread in both cases. "Another" implies that isn't the case.

> If a thread in {es} wants to send data over a particular channel then it passes
> the data to the owning `transport_worker` thread for the actual transmission.

Reviewer: ```suggestion
If a thread in {es} wants to send data over a particular channel, it passes
the data to the owning `transport_worker` thread for the actual transmission.
```

Normally the `transport_worker` threads will not completely handle the messages
they receive. Instead, they will do a small amount of preliminary processing
and then dispatch (hand off) the message to a different
<<modules-threadpool,threadpool>> for the rest of their handling. For instance,
bulk messages are dispatched to the `write` threadpool, searches are dispatched
to one of the `search` threadpools, and requests for statistics and other
management tasks are mostly dispatched to the `management` threadpool. However
in some cases the processing of a message is expected to be so quick that {es}
will do all of the processing on the `transport_worker` thread rather than
incur the overhead of dispatching it elsewhere.

> There are relatively few `transport_worker` threads, by default one per CPU,
> and many more TCP channels, possibly tens-of-thousands or more. If data arrives
> on a TCP channel and its owning `transport_worker` thread is busy then the data
> will not be processed until the thread finishes whatever it is doing.
> Similarly, outgoing data are not actually sent over a channel until the owning
> `transport_worker` thread is free. This means that we require every
> `transport_worker` thread to be idle frequently, which looks something like
> this in a stack dump:

Reviewer: ```suggestion
By default, there is one `transport_worker` thread per CPU. In contrast, there
are tens-of-thousands or more TCP channels. If data arrives
on a TCP channel and its owning `transport_worker` thread is busy, the data
isn't processed until the thread finishes whatever it is doing.
Similarly, outgoing data are not actually sent over a channel until the owning
`transport_worker` thread is free. This means that we require every
`transport_worker` thread to be idle frequently, which looks something like
this in a stack dump:
```

```
"elasticsearch[instance-0000000000][transport_worker][T#1]" #28 daemon prio=5 os_prio=0 cpu=1232378.30ms elapsed=47152.02s tid=0x00007fc8a8569da0 nid=0x16f runnable  [0x00007fc8ac5e6000]
   java.lang.Thread.State: RUNNABLE
        at sun.nio.ch.EPoll.wait(java.base@15.0.1/Native Method)
        at sun.nio.ch.EPollSelectorImpl.doSelect(java.base@15.0.1/EPollSelectorImpl.java:120)
        at sun.nio.ch.SelectorImpl.lockAndDoSelect(java.base@15.0.1/SelectorImpl.java:129)
        - locked <0x0000000624849c60> (a sun.nio.ch.Util$2)
        - locked <0x0000000624849c08> (a sun.nio.ch.EPollSelectorImpl)
        at sun.nio.ch.SelectorImpl.select(java.base@15.0.1/SelectorImpl.java:146)
        at io.netty.channel.nio.NioEventLoop.select(NioEventLoop.java:803)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:457)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at java.lang.Thread.run(java.base@15.0.1/Thread.java:832)
```

> In the <<cluster-nodes-hot-threads>> API an idle `transport_worker` thread will
> be reported like this:

Reviewer: ```suggestion
In the <<cluster-nodes-hot-threads>> API an idle `transport_worker` thread is
reported like this:
```

```
100.0% [cpu=0.0%, other=100.0%] (500ms out of 500ms) cpu usage by thread 'elasticsearch[instance-0000000000][transport_worker][T#1]'
     10/10 snapshots sharing following 9 elements
       java.base@17.0.1/sun.nio.ch.EPoll.wait(Native Method)
       java.base@17.0.1/sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:118)
       java.base@17.0.1/sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:129)
       java.base@17.0.1/sun.nio.ch.SelectorImpl.select(SelectorImpl.java:146)
       io.netty.channel.nio.NioEventLoop.select(NioEventLoop.java:810)
       io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:457)
       io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:986)
       io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
       java.base@17.0.1/java.lang.Thread.run(Thread.java:833)
```

> Note that `transport_worker` threads should always be in state `RUNNABLE`,
> because they block in the native `EPoll#wait` method, so the hot threads API
> will report them at 100% overall utilisation. This is normal, and the breakdown
> of time into `cpu=` and `other=` fractions shows how much time the thread spent
> running and waiting for input respectively.

Reviewer: ```suggestion
Note that `transport_worker` threads should always be in state `RUNNABLE`,
because they block in the native `EPoll#wait` method. The hot threads API
will report these threads at 100% overall utilisation. This is normal, and the breakdown
of time into `cpu=` and `other=` fractions shows how much time the thread spent
running and waiting for input respectively.
```

> If a `transport_worker` thread is not frequently idle then it will build up a
> backlog of work, which will cause delays in processing messages on the channels
> that it owns. It is hard to predict exactly which work will be delayed:

Reviewer: ```suggestion
If a `transport_worker` thread is not frequently idle, it will build up a
backlog of work. This can cause delays in processing messages on the channels
that it owns. It's hard to predict exactly which work will be delayed:
```

* There are many more channels than threads. If work related to one channel is
causing delays to its worker thread then all other channels owned by that
thread will also suffer delays.

* The mapping from TCP channels to worker threads is fixed, but arbitrary. Each
channel is assigned an owning thread in a round-robin fashion when the channel
is opened. Each worker thread is responsible for many different kinds of
channel.

* There are many channels open between each pair of nodes, and for each request
{es} will choose from the appropriate channels in a round-robin fashion. Some
requests may end up on a channel owned by a delayed worker while other
identical requests will be sent on a channel that's working smoothly.

> If the backlog builds up too far then some messages may be delayed by many
> seconds and the node might even <<cluster-fault-detection,fail its health
> checks>> and be removed from the cluster. Sometimes you can find evidence of
> busy `transport_worker` threads using the <<cluster-nodes-hot-threads>> API,
> but this API itself sends network messages so may not work correctly if the
> `transport_worker` threads are too busy. It is more reliable to use `jstack` to
> obtain stack dumps, or Java Flight Recorder to obtain a profiling trace, since
> these tools are independent of any work the JVM is performing.

Reviewer: ```suggestion
If the backlog builds up too far, some messages may be delayed by many
seconds. The node might even <<cluster-fault-detection,fail its health
checks>> and be removed from the cluster. Sometimes, you can find evidence of
busy `transport_worker` threads using the <<cluster-nodes-hot-threads>> API.
However, this API itself sends network messages so may not work correctly if the
`transport_worker` threads are too busy. It is more reliable to use `jstack` to
obtain stack dumps or use Java Flight Recorder to obtain a profiling trace.
These tools are independent of any work the JVM is performing.
```
